src/parser/utils.ts

```diff
@@ -1,65 +1,55 @@
-const map = new WeakMap<object, string>();
-
-const TYPES = {
-  ArrayBuffer: '[object ArrayBuffer]',
-  SharedArrayBuffer: '[object SharedArrayBuffer]',
-  Uint8Array: '[object Uint8Array]',
-  BigInt64Array: '[object BigInt64Array]',
-  BigUint64Array: '[object BigUint64Array]',
-  RegExp: '[object RegExp]',
-  Map: '[object Map]',
-  Date: '[object Date]'
-};
-
-/**
- * Retrieves the prototype.toString() of a value.
- * If the value is an object, it will cache the result in a WeakMap for future use.
- */
-function getPrototypeString(value: unknown): string {
-  let str = map.get(value as object);
-
-  if (!str) {
-    str = Object.prototype.toString.call(value);
-    if (value !== null && typeof value === 'object') {
-      map.set(value, str);
-    }
-  }
-  return str;
-}
-
 export function isAnyArrayBuffer(value: unknown): value is ArrayBuffer {
-  const type = getPrototypeString(value);
-  return type === TYPES.ArrayBuffer || type === TYPES.SharedArrayBuffer;
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    Symbol.toStringTag in value &&
+    (value[Symbol.toStringTag] === 'ArrayBuffer' ||
+      value[Symbol.toStringTag] === 'SharedArrayBuffer')
+  );
 }
 
 export function isUint8Array(value: unknown): value is Uint8Array {
-  const type = getPrototypeString(value);
-  return type === TYPES.Uint8Array;
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    Symbol.toStringTag in value &&
+    value[Symbol.toStringTag] === 'Uint8Array'
+  );
 }
 
 export function isBigInt64Array(value: unknown): value is BigInt64Array {
-  const type = getPrototypeString(value);
-  return type === TYPES.BigInt64Array;
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    Symbol.toStringTag in value &&
+    value[Symbol.toStringTag] === 'BigInt64Array'
+  );
 }
 
 export function isBigUInt64Array(value: unknown): value is BigUint64Array {
-  const type = getPrototypeString(value);
-  return type === TYPES.BigUint64Array;
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    Symbol.toStringTag in value &&
+    value[Symbol.toStringTag] === 'BigUint64Array'
+  );
 }
 
 export function isRegExp(d: unknown): d is RegExp {
-  const type = getPrototypeString(d);
-  return type === TYPES.RegExp;
+  return Object.prototype.toString.call(d) === '[object RegExp]';
 }
 
-export function isMap(d: unknown): d is Map<unknown, unknown> {
-  const type = getPrototypeString(d);
-  return type === TYPES.Map;
+export function isMap(value: unknown): value is Map<unknown, unknown> {
+  return (
+    typeof value === 'object' &&
+    value != null &&
+    Symbol.toStringTag in value &&
+    value[Symbol.toStringTag] === 'Map'
+  );
 }
 
 export function isDate(d: unknown): d is Date {
-  const type = getPrototypeString(d);
-  return type === TYPES.Date;
+  return Object.prototype.toString.call(d) === '[object Date]';
```

Comment:
Does it make sense to preface this check with an `d instanceof Date` check?

Reply:
Possibly I'm checking the speed of the following now:

```js
const isDateInstance = (date) => date instanceof Date;
const isDateString = (date) => Object.prototype.toString.call(date) === '[object Date]';
const isDateBrand = (date) => {
	try {
		Date.prototype.getSeconds.call(date)
		return true;
	} catch {
		return false;
	}
}

const goodInput = new Date();
const badInput = (new Date()).toString();
```

the try-catch one seems really fast on success 👀 but terrible on failure, I'll see if instanceof speeds up over the string check

Reply:
<img width="689" alt="image" src="https://github.com/user-attachments/assets/ac68de05-8592-4d85-9cad-cc486535ec96">


I'll add an `instanceof` and `||` 🙂